csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs

```diff
@@ -0,0 +1,1211 @@
+        }
+    }
+
+    /// <summary>
+    /// Asynchronously retrieves flight information for a given flight descriptor.
+    /// </summary>
+    /// <param name="options">RPC-layer hints for this call.</param>
+    /// <param name="descriptor">The descriptor of the dataset request, whether a named dataset or a command.</param>
+    /// <returns>A task that represents the asynchronous operation. The task result contains the FlightInfo describing where to access the dataset.</returns>
+    public async Task<FlightInfo> GetFlightInfoAsync(FlightCallOptions options, FlightDescriptor descriptor)
+    {
+        if (descriptor is null)
+        {
+            throw new ArgumentNullException(nameof(descriptor));
+        }
+
+        try
+        {
+            var flightInfoCall = _client.GetInfo(descriptor, options.Headers);
+            var flightInfo = await flightInfoCall.ResponseAsync.ConfigureAwait(false);
+            return flightInfo;
+        }
+        catch (RpcException ex)
+        {
+            throw new InvalidOperationException("Failed to get flight info", ex);
+        }
+    }
+
+    /// <summary>
+    /// Asynchronously retrieves flight information for a given flight descriptor.
+    /// </summary>
+    /// <param name="descriptor">The descriptor of the dataset request, whether a named dataset or a command.</param>
+    /// <returns>A task that represents the asynchronous operation. The task result contains the FlightInfo describing where to access the dataset.</returns>
+    public Task<FlightInfo> GetFlightInfoAsync(FlightDescriptor descriptor)
+    {
+        var options = new FlightCallOptions();
+        return GetFlightInfoAsync(options, descriptor);
+    }
+
+    /// <summary>
+    /// Perform the indicated action, returning an iterator to the stream of results, if any.
+    /// </summary>
+    /// <param name="options">Per-RPC options</param>
+    /// <param name="action">The action to be performed</param>
+    /// <returns>An async enumerable of results</returns>
+    public async IAsyncEnumerable<FlightResult> DoActionAsync(FlightCallOptions options, FlightAction action)
+    {
+        if (options is null)
+            throw new ArgumentNullException(nameof(options));
+
+        if (action is null)
+            throw new ArgumentNullException(nameof(action));
+
+        var call = _client.DoAction(action, options.Headers);
+
+        await foreach (var result in call.ResponseStream.ReadAllAsync())
+        {
+            yield return result;
+        }
+    }
+
+    /// <summary>
+    /// Perform the indicated action with default options, returning an iterator to the stream of results, if any.
+    /// </summary>
+    /// <param name="action">The action to be performed</param>
+    /// <returns>An async enumerable of results</returns>
+    public async IAsyncEnumerable<FlightResult> DoActionAsync(FlightAction action)
+    {
+        await foreach (var result in DoActionAsync(new FlightCallOptions(), action))
```

Comment:
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

Reply:
done